### DISCUSSION TITLE
Sentry <> Starlette: Error Tracking and Performance Monitoring

### DISCUSSION THREAD

**Discussion by smeubank** [encode/starlette#1787] [Q&A]
### Announcement:

Sentry has delivered a dedicated integration for Starlette! 

### What is available today?

- The integration is available with manual detection today - and will be automatic in the near future
- Check our docs here: [Getting Started](https://docs.sentry.io/platforms/python/guides/starlette/)
- Project onboarding experience:

<img width="1412" alt="image" src="https://user-images.githubusercontent.com/47563310/182342503-783b8852-5c27-4b5d-b8e4-e1dd87544e1b.png">

### What we would like to feedback on?

Any and all feedback, issues, PRs are welcome! We will respond in a timely manner on the [sentry-python SDK repo](https://github.com/getsentry/sentry-python/issues)

- What are your use cases, that are most important to you?
- Is there specific context for error tracking or performance monitoring that you would like for us to cover?
- Spans for all the middleware is available for Starlette. Can you confirm how this is working for you?

![image](https://user-images.githubusercontent.com/47563310/182342121-3a0b29bf-78ab-45a7-8762-98b5adbb8d92.png)

**Comment by erewok** [comment]
This looks promising: we use Sentry at my work and I am a fan of their services. 

We have written our own Prometheus-middleware for tracking latencies and request success/failure rates, and I had a question about your comment that:

> Spans for all the middleware is available for Starlette

Does this automatically include performance information (I have seen some mention of this in general from Sentry recently) or response times or other interesting data like that? Does Sentry also offer waterfall graphs (a la Jaeger or other opentelemetry tooling) and is that what you meant by spans here? 

Thanks again!

**Reply by smeubank** [reply]
Hi @erewok 

Spans are sent in an array, inside transaction events payloads, which power the performance product. You can check out our SDK [developer docs on transaction payloads](https://develop.sentry.dev/sdk/event-payloads/transaction/), if you want to go farther down the rabbit hole. 

But simply, yes this includes performance information. I think in this case we are talking about [distributed tracing](https://docs.sentry.io/product/sentry-basics/tracing/distributed-tracing/), in which case we have traces, made up of transactions, which are made up of span(s)

The docs are meaty, but basically there are some graphs there which tell the story simply.

![image](https://user-images.githubusercontent.com/47563310/182913445-c47c1928-27a8-419d-aefb-1a878705463f.png)

And yes, specifically sentry offers [trace view](https://docs.sentry.io/product/sentry-basics/tracing/trace-view/) with waterfall graphs within the Performance product.

![image](https://user-images.githubusercontent.com/47563310/182912815-730af57d-58f5-4e7f-9952-e4822ebae20c.png)

If you want any more help to onboard to our Performance product please let me know!

**Reply by erewok** [reply]
That looks incredible! Thanks for the information.

**Comment by lovelydinosaur** [comment]
Fantastic work. 🌟

One actionable item, and one item for consideration here..

* A pull request adding this to the third party docs would help highlight the package... https://www.starlette.io/third-party-packages/
* I also wonder if it'd make sense to include a documentation change on https://www.starlette.io/exceptions/ - we don't typically have third party packages listed in those sections, but sentry's tooling here provides *such* an essential part of production deployments, that I think it *could* be valid to highlight it prominently in this way.

**Reply by smeubank** [reply]
Both are great suggestion! Thank you @tomchristie sharing with the team now

**Reply by antonpirker** [reply]
* The PR for third party docs is a good idea. I will create one after my vacation (in three weeks)
* And mentioning the sentry integration in the exceptions page would be very cool! Could we add a link to our setup guide [1] there? 

1: https://docs.sentry.io/platforms/python/guides/starlette/

**Reply by lovelydinosaur** [reply]
> Could we add a link to our setup guide there?

I'd suggest we determine together as a team against an individual pull request what we consider appropriate.

*(It seems like a decent idea to me.)*

**Reply by smeubank** [reply]
Anton and I are both on holidays at the moment. Just looping in @sl0thentr0py

Adding mentions and possible links to the pages in starlette sounds agreed. We should probably also ensure in our sentry docs if we are missing context &/or links to help starlette devs back to any specific locations in your docs. (Now I'm confusing myself on the back and forth of docs 😅)

**Reply by smeubank** [reply]
@antonpirker fyi, happy for your input on wording and links. For correctness and relevance

https://github.com/encode/starlette/pull/2063

https://github.com/encode/starlette/pull/2062

**Comment by Kludex** [comment]
Not sure if it's relevant, but on `WebSocket`'s you'll have always two exceptions being raised:

```python
import os

import sentry_sdk
from sentry_sdk.integrations.starlette import StarletteIntegration
from starlette.applications import Starlette
from starlette.routing import WebSocketRoute
from starlette.websockets import WebSocket

SENTRY_DSN = os.getenv("SENTRY_DSN")

sentry_sdk.init(
    dsn=SENTRY_DSN,
    integrations=[StarletteIntegration()],
    # Set traces_sample_rate to 1.0 to capture 100%
    # of transactions for performance monitoring.
    # We recommend adjusting this value in production,
    traces_sample_rate=1.0,
)

class CustomWebSocketException(Exception):
    ...


async def websocket_error(websocket: WebSocket):
    await websocket.accept()
    raise CustomWebSocketException()


app = Starlette(routes=[WebSocketRoute("/sentry-ws", websocket_error)])
```

![image](https://user-images.githubusercontent.com/7353520/184530617-24258abe-3bc0-40dc-bb96-d3077db30a43.png)

**Comment by Kludex** [comment]
Again, not sure if it's relevant... But the `WebSocket`s transaction is using the `endpoint` style, even when "url" is set.

![image](https://user-images.githubusercontent.com/7353520/184530814-dd5453f7-d98c-43c9-822b-ddcf37ba2a07.png)

**Comment by Kludex** [comment]
As a note... If there's an error handler (500 or Exception) on the application, the package still sends the error to Sentry. This is great!

This is not the case on Flask when handling `Exception` (it's when handling 500 tho)! :eyes:

**Comment by Kludex** [comment]
The "App Only" option doesn't work as expected, I guess? I see the full trace from Starlette. I would expect only the `main.py` being presented there. :thinking: 

![image](https://user-images.githubusercontent.com/7353520/184531089-d7c95360-d506-42e0-8d31-cc3371cc440c.png)

**Comment by Kludex** [comment]
That's what I could find on the Error Tracking. The application used: https://github.com/Kludex/starlette-sentry. :pray: 

My trial expires in three days, can I have a free plan to do further testing on performance? :eyes:

**Reply by antonpirker** [reply]
Hey @Kludex! Thanks for the amazing testing work!
There is the "Developer" account [1] that has some free quote (I guess 50k errors a month) have you already depleted this?

1: https://sentry.io/pricing/

**Reply by Kludex** [reply]
I don't know how to switch to that... "Your free trial has ended. You may [contact support](mailto:support@sentry.io?subject=Request%20Another%20Trial) to request another trial."

**Reply by antonpirker** [reply]
Can you give me your username or email you use for sentry? (sent to anton dot pirker at sentry io ) Then I can ask around to get you more quota

**Comment by antonpirker** [comment]
We had a bug report [1] about an app that has a Starlette middleware and Sentry enabled and when trying to read the request body in the middleware it hangs because Sentry already read the body. 

What is happening is described in this commit:
https://github.com/getsentry/sentry-python/issues/1573#issuecomment-1230237904

My question now is: 
Why is `BaseHTTPMiddleware` not using the starlette `Request` object that it has to read the body but uses anyio directly?

Using the Starlette `Request` would cache the read data from the body stream and would make it possible to read it multiple times. 

1: https://github.com/getsentry/sentry-python/issues/1573

**Reply by antonpirker** [reply]
maybe @Kludex or @tomchristie can help me understand?

**Reply by antonpirker** [reply]
Just found this discussion that talks about the limitation sof BaseHTTPMiddleware, so I guess I will find my answer there: 
https://github.com/encode/starlette/discussions/1729
